sources/subs/Permissions.class.php

```diff
@@ -0,0 +1,201 @@
+				' . (empty($parents) ? '' : 'AND id_parent IN ({array_int:parent_list})'),
+			array(
+				'parent_list' => $parents,
+				'not_inherited' => -2,
+			)
+		);
+		$children = array();
+		$parents = array();
+		$child_groups = array();
+		while ($row = $db->fetch_assoc($request))
+		{
+			$children[$row['id_parent']][] = $row['id_group'];
+			$child_groups[] = $row['id_group'];
+			$parents[] = $row['id_parent'];
+		}
+		$db->free_result($request);
+
+		$parents = array_unique($parents);
+
+		// Not a sausage, or a child?
+		if (empty($children))
+			return false;
+
+		// First off, are we doing general permissions?
+		if ($profile < 1 || $profile === null)
+		{
+			// Fetch all the parent permissions.
+			$request = $db->query('', '
+				SELECT id_group, permission, add_deny
+				FROM {db_prefix}permissions
+				WHERE id_group IN ({array_int:parent_list})',
+				array(
+					'parent_list' => $parents,
+				)
+			);
+			$permissions = array();
+			while ($row = $db->fetch_assoc($request))
+			{
+				foreach ($children[$row['id_group']] as $child)
+				{
+					$permissions[] = array('id_group' => (int) $child, 'permission' => $row['permission'], 'add_deny' => $row['add_deny']);
+				}
+			}
+			$db->free_result($request);
+
+			$db->query('', '
+				DELETE FROM {db_prefix}permissions
+				WHERE id_group IN ({array_int:child_groups})',
+				array(
+					'child_groups' => $child_groups,
+				)
+			);
+
+			// Finally insert.
+			if (!empty($permissions))
+			{
+				require_once(SUBSDIR . '/ManagePermissions.subs.php');
+				replacePermission($permissions);
+			}
+		}
+
+		// Then, what about board profiles?
+		if ($profile != -1)
+		{
+			$profileQuery = $profile === null ? '' : ' AND id_profile = {int:current_profile}';
+
+			// Again, get all the parent permissions.
+			$request = $db->query('', '
+				SELECT id_profile, id_group, permission, add_deny
+				FROM {db_prefix}board_permissions
+				WHERE id_group IN ({array_int:parent_groups})
+					' . $profileQuery,
+				array(
+					'parent_groups' => $parents,
+					'current_profile' => $profile !== null && $profile ? $profile : 1,
+				)
+			);
+			$permissions = array();
+			while ($row = $db->fetch_assoc($request))
+				foreach ($children[$row['id_group']] as $child)
+					$permissions[] = array($child, $row['id_profile'], $row['permission'], $row['add_deny']);
+			$db->free_result($request);
+
+			$db->query('', '
+				DELETE FROM {db_prefix}board_permissions
+				WHERE id_group IN ({array_int:child_groups})
+					' . $profileQuery,
+				array(
+					'child_groups' => $child_groups,
+					'current_profile' => $profile !== null && $profile ? $profile : 1,
+				)
+			);
+
+			// Do the insert.
+			if (!empty($permissions))
+			{
+				$db->insert('insert',
+					'{db_prefix}board_permissions',
+					array('id_group' => 'int', 'id_profile' => 'int', 'permission' => 'string', 'add_deny' => 'int'),
+					$permissions,
+					array('id_group', 'id_profile', 'permission')
+				);
+			}
```

Comment:
Another one.

			$permissions = array();
			while ($row = $db->fetch_assoc($request))
				foreach ($children[$row['id_group']] as $child)
					$permissions[] = array($row['permission'], $child, $row['add_deny'], $row['id_profile']);
			$db->free_result($request);

			// Do the insert.
			if (!empty($permissions))
			{
				replaceBoardPermission($permissions);
			}